requests-for-comments/2025-04-15-scripts-to-rule-them-all.md

```diff
@@ -0,0 +1,72 @@
+lot of friction as we switch from project to project and have to learn a whole new set of incantations. Not only that, it makes it more challenging
+for external contributors to get involved as they have to learn a whole new process for each repository they work on.
+
+And as we continue to grow, this challenge will only grow.
+
+## Solution
+
+Adopt and adapt GitHub's approach of [scripts to rule them all](https://github.blog/engineering/engineering-principles/scripts-to-rule-them-all/) to
+fit our needs.
+
+> Being able to get from git clone to an up-and-running project in a development environment is imperative for fast, reliable contributions.
+
+I'm not suggesting we do _exactly_ what they do, but a similar approach would be a big benefit. Not every repository will need every script. Some
+repositories will need scripts custom to the environment. That's all fine. The goal is to have a baseline set of scripts that we can use to get a
+project up and running in a development environment and a known location to look for those scripts.
+
+## Context
+
+For context, here's the list of scripts referenced in the GitHub blog post ([samples available in a repository](https://github.com/github/scripts-to-rule-them-all)):
+
+* `script/bootstrap` - Install dependencies (Ex. npm packages, brew packages, etc. Usually run once after cloning the repository)
+* `script/setup` - Setup the project. Seems like this could just be merged in bootstrap
+* `script/update` - Update project after a pull such as migrations, etc.
+* `script/server` - For web apps, starts the application. For client SDKs, might start an example site.
+* `script/test` - Run tests (Ex. `npm test`, `bundle exec rspec`, etc.). Also includes linting, formatting, etc.
+* `script/cibuild` - used by the CI system to run tests, build, etc.
+* `script/console` - is used to open a console for your application.
+
+## Design
+
+For `posthog/posthog`, we already have a set of scripts in the `bin` directory. One issue with using `bin` across the board is that some environments
+add `bin` to the `.gitignore` file by default because that's where they compile binaries to. However, we've already set this precedence, so it may make
+sense to stick to it. For example, [posthog/posthog-dotnet](https://github.com/PostHog/posthog-dotnet/tree/main/bin) already has a `bin` directory.
+
+* `bin/setup` - Install or upgrade dependencies (Ex. npm packages, brew packages, etc. Usually run once after cloning the repository).
+* `bin/update` - Updates dependencies after a pull. This could simply call `bin/setup`.
+* `bin/build` - Build the project, for projects that are compiled such as C#, Java, etc.
+* `bin/start` - Start the project. For SDKs, this might start an example server.
+* `bin/test` - Run tests (Ex. `npm test`, `bundle exec rspec`, etc.). Also includes linting, formatting, etc.
+* `bin/fmt` - Optional: Format/lint code. This can be called by `test`.
```

Comment:
I think `fmt` and `lint` should be different tasks

Reply:
Maybe `lint` should just go into the `bin/build` task? Hear me out.

It seems to me that linting for a dynamic language serves a similar purpose to compilation in a compiled language in that it points out obvious errors.

So for a language like python, `bin/build` could run both `mypy` and `flake8`. Thoughts?

Reply:
I'm open to just having it be a separate script as well.

Reply:
I think for speedy, i prefer the `lint` to be its own thing, but it really depends on the tooling.
what i did in the past and worked really well is that i have a cmd for each step (fmt, lint, etc), and a cmd that does it all (eg build) that will run all its dependencies (fmt lint, etc).
eg https://github.com/PostHog/posthog-android/blob/9906ac8ac333a0e4990b64dceeca892453d81303/Makefile#L6-L7
^ runs everything else, so in CI I just do that, but locally if i want to check something, i can just `make format`